### ISSUE TITLE
Investigate generation of functions which have outputs that depend on their inputs

### ISSUE THREAD

**Issue by LightAndLight** [qfpl/hedgehog-fn#3]
There were a couple of questions in [the reddit thread](https://www.reddit.com/r/haskell/comments/956knr/hedgehogfn_function_generation_for_hedgehog/) about this kind of idea.

I think it's cool but I don't know if it's possible. The main example of monotonicity requires knowledge of the entire domain and codomain, which isn't captured by `a -> Gen b` anyway. At the very least we need to be aware of the history of inputs to the function.

The memo-trie approach works by lazily enumerating all the possible inputs to the function, so if we wanted to generate a monotonic function then somehow each input would have to have a knot that knows about all the other inputs. I don't know that it works.

**Comment by tscholak** [comment]
lol, I just ran into this problem. any idea how this may be accomplished?

**Comment by LightAndLight** [comment]
The monotonic example, or something else? I haven't thought about it for a while. What are you trying to do?

**Comment by tscholak** [comment]
I wanted to generate a function isomorphic to `a -> Either a b`.

**Comment by tscholak** [comment]
In the meantime, I circumvented the issue by generating `Either () (Fn a b)` and mapping this to `a -> Either a b` down the line. In the `Left` case, it's just the identity.

**Comment by LightAndLight** [comment]
Do you mean you generated an `a -> Either a b` for some specific `a` and `b` (say, `Int` and `Bool`)? Or you want to generate a parametric version (`forall a b. a -> Either a b` - there's only one function of this type- `Left`)?

**Comment by tscholak** [comment]
no, no, you are right, it would be for some unknown, but specific `a` and `b`.

implementing the following is not possible with what `Hedgehog.Function` provides today:
```
mkGen
  :: forall a b
   . (Hedgehog.Function.Arg a, Hedgehog.Function.Vary a)
  => Hedgehog.Gen b
  -> Hedgehog.Gen (Hedgehog.Function.Fn a (Either a b))
```
